lib/widgets/action_sheet.dart

```diff
@@ -1,36 +1,182 @@
+///
+/// Must have a [MessageListPage] ancestor.
 void showMessageActionSheet({required BuildContext context, required Message message}) {
+  // The UI that's conditioned on this won't live-update during this appearance
+  // of the action sheet (we avoid calling composeBoxControllerOf in a build
+  // method; see its doc). But currently it will be constant through the life of
+  // any message list, so that's fine.
+  final isComposeBoxOffered = MessageListPageState.composeBoxControllerOf(context) != null;
   showDraggableScrollableModalBottomSheet(
     context: context,
-    builder: (BuildContext context) {
+    builder: (BuildContext innerContext) {
       return Column(children: [
-        MenuItemButton(
-          leadingIcon: Icon(Icons.adaptive.share),
-          onPressed: () async {
-            // Close the message action sheet; we're about to show the share
-            // sheet. (We could do this after the sharing Future settles, but
-            // on iOS I get impatient with how slowly our action sheet
-            // dismisses in that case.)
-            // TODO(#24): Fix iOS bug where this call causes the keyboard to
-            //   reopen (if it was open at the time of this
-            //   `showMessageActionSheet` call) and cover a large part of the
-            //   share sheet.
-            Navigator.of(context).pop();
-
-            // TODO: to support iPads, we're asked to give a
-            //   `sharePositionOrigin` param, or risk crashing / hanging:
-            //     https://pub.dev/packages/share_plus#ipad
-            //   Perhaps a wart in the API; discussion:
-            //     https://github.com/zulip/zulip-flutter/pull/12#discussion_r1130146231
-            // TODO: Share raw Markdown, not HTML
-            await Share.shareWithResult(message.content);
-          },
-          child: const Text('Share')),
+        ShareButton(message: message, bottomSheetContext: innerContext, messageListContext: context),
+        if (isComposeBoxOffered) QuoteAndReplyButton(
+          message: message,
+          bottomSheetContext: innerContext,
+          messageListContext: context,
+        ),
       ]);
     });
 }
+
+abstract class MessageActionSheetMenuItemButton extends StatelessWidget {
+  MessageActionSheetMenuItemButton({
+    super.key,
+    required this.message,
+    required this.bottomSheetContext,
+    required this.messageListContext,
+  }) : assert(bottomSheetContext.findAncestorWidgetOfExactType<BottomSheet>() != null),
+       assert(messageListContext.findAncestorWidgetOfExactType<MessageListPage>() != null);
+
+  IconData get icon;
+  String get label;
+  VoidCallback get onPressed;
+
+  final Message message;
+  final BuildContext bottomSheetContext;
+  final BuildContext messageListContext;
+
+  @override
+  Widget build(BuildContext context) {
+    return MenuItemButton(
+      leadingIcon: Icon(icon),
+      onPressed: onPressed,
```

Comment:
```suggestion
      onPressed: () => onPressed(context),
```

Then I think `bottomSheetContext` becomes unnecessary; the button's own context is just as good for everything these widgets are doing with it.